sensing/autoware_pointcloud_preprocessor/src/pointcloud_densifier/pointcloud_densifier_node.cpp

```diff
@@ -0,0 +1,331 @@
+  size_t output_size = 0;
+
+  if (indices && !indices->empty()) {
+    for (const auto & idx : *indices) {
+      const size_t data_offset = idx * input_cloud->point_step;
+
+      float x, y;
+      std::memcpy(&x, &input_cloud->data[data_offset + x_offset], sizeof(float));
+      std::memcpy(&y, &input_cloud->data[data_offset + y_offset], sizeof(float));
+
+      if (x > param_.x_min && x < param_.x_max && y > param_.y_min && y < param_.y_max) {
+        // Copy point data if within ROI
+        std::memcpy(
+          &filtered_cloud->data[output_size], &input_cloud->data[data_offset],
+          input_cloud->point_step);
+        output_size += input_cloud->point_step;
+      }
+    }
+  } else {
+    for (size_t i = 0; i < input_cloud->width * input_cloud->height; ++i) {
+      const size_t data_offset = i * input_cloud->point_step;
+
+      // Get x,y coordinates
+      float x, y;
+      std::memcpy(&x, &input_cloud->data[data_offset + x_offset], sizeof(float));
+      std::memcpy(&y, &input_cloud->data[data_offset + y_offset], sizeof(float));
+
+      if (x > param_.x_min && x < param_.x_max && y > param_.y_min && y < param_.y_max) {
+        // Copy point data if within ROI
+        std::memcpy(
+          &filtered_cloud->data[output_size], &input_cloud->data[data_offset],
+          input_cloud->point_step);
+        output_size += input_cloud->point_step;
+      }
+    }
+  }
+
+  // Set width based on actual number of points
+  filtered_cloud->width = output_size / filtered_cloud->point_step;
+  filtered_cloud->row_step = filtered_cloud->width * filtered_cloud->point_step;
+  filtered_cloud->data.resize(output_size);
+
+  return filtered_cloud;
+}
+
+void PointCloudDensifierNode::transformAndMergePreviousClouds(
+  const PointCloud2ConstPtr & current_msg, const OccupancyGrid & occupancy_grid,
+  PointCloud2 & combined_cloud)
+{
+  int x_offset = current_msg->fields[pcl::getFieldIndex(*current_msg, "x")].offset;
+  int y_offset = current_msg->fields[pcl::getFieldIndex(*current_msg, "y")].offset;
+
+  for (const auto & previous_cloud : previous_pointclouds_) {
+    if (!previous_cloud || previous_cloud->data.empty()) {
+      continue;
+    }
+
+    geometry_msgs::msg::TransformStamped transform_stamped;
+    try {
+      tf2::TimePoint current_time_point = tf2::TimePoint(
+        std::chrono::nanoseconds(current_msg->header.stamp.nanosec) +
+        std::chrono::seconds(current_msg->header.stamp.sec));
+      tf2::TimePoint prev_time_point = tf2::TimePoint(
+        std::chrono::nanoseconds(previous_cloud->header.stamp.nanosec) +
+        std::chrono::seconds(previous_cloud->header.stamp.sec));
+      transform_stamped = tf_buffer_->lookupTransform(
+        current_msg->header.frame_id, current_time_point, previous_cloud->header.frame_id,
+        prev_time_point, "map");
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_WARN(get_logger(), "Could not transform point cloud: %s", ex.what());
+      continue;
+    }
```

Comment:
[imho]
By maintaining point clouds in the map frame and setting the occupancy grid's origin relative to self-localization, we could reduce the number of transforms required for each point cloud, leading to a simpler and more economical design.

Reply:
The reason I prefer this design is that I don't want to transform the point cloud twice in a row, base_link -> map -> base_link .  The proposed solution only transforms base_link -> base_link.